### ISSUE TITLE
Missing symbols issue with Anaconda installation on Mac

### ISSUE THREAD

**Issue by yoavartzi** [clab/dynet#373]
Did anyone encounter installation issues with Anaconda on Mac? 

I am getting `Undefined symbols for architecture x86_64` issues during `make -j 2`. `cmake` works fine. 

```
Undefined symbols for architecture x86_64:
  "boost::archive::text_iarchive_impl<boost::archive::text_iarchive>::load(std::basic_string<char, std::char_traits<char>, std::allocator<char> >&)", referenced from:
      boost::archive::detail::iserializer<boost::archive::text_iarchive, std::pair<std::basic_string<char, std::char_traits<char>, std::allocator<char> > const, int> >::load_object_data(boost::archive::detail::basic_iarchive&, void*, unsigned int) const in dict.cc.o
      boost::archive::detail::iserializer<boost::archive::text_iarchive, std::vector<std::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >::load_object_data(boost::archive::detail::basic_iarchive&, void*, unsigned int) const in dict.cc.o
  "boost::archive::text_iarchive_impl<boost::archive::text_iarchive>::text_iarchive_impl(std::basic_istream<char, std::char_traits<char> >&, unsigned int)", referenced from:
      dynet::load_dynet_model(std::basic_string<char, std::char_traits<char>, std::allocator<char> >, dynet::Model*) in model.cc.o
  "boost::archive::text_oarchive_impl<boost::archive::text_oarchive>::save(std::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)", referenced from:
      boost::archive::detail::oserializer<boost::archive::text_oarchive, std::pair<std::basic_string<char, std::char_traits<char>, std::allocator<char> > const, int> >::save_object_data(boost::archive::detail::basic_oarchive&, void const*) const in dict.cc.o
      boost::archive::detail::oserializer<boost::archive::text_oarchive, std::vector<std::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >::save_object_data(boost::archive::detail::basic_oarchive&, void const*) const in dict.cc.o
      boost::archive::detail::common_oarchive<boost::archive::text_oarchive>::vsave(boost::archive::class_name_type const&) in model.cc.o
  "boost::archive::text_oarchive_impl<boost::archive::text_oarchive>::text_oarchive_impl(std::basic_ostream<char, std::char_traits<char> >&, unsigned int)", referenced from:
      dynet::save_dynet_model(std::basic_string<char, std::char_traits<char>, std::allocator<char> >, dynet::Model*) in model.cc.o
  "boost::archive::basic_text_iprimitive<std::basic_istream<char, std::char_traits<char> > >::~basic_text_iprimitive()", referenced from:
      dynet::load_dynet_model(std::basic_string<char, std::char_traits<char>, std::allocator<char> >, dynet::Model*) in model.cc.o
      boost::archive::text_iarchive_impl<boost::archive::text_iarchive>::~text_iarchive_impl() in model.cc.o
      boost::archive::text_iarchive_impl<boost::archive::text_iarchive>::~text_iarchive_impl() in model.cc.o
      boost::archive::text_iarchive::~text_iarchive() in model.cc.o
      boost::archive::text_iarchive::~text_iarchive() in model.cc.o
  "boost::archive::basic_text_oprimitive<std::basic_ostream<char, std::char_traits<char> > >::~basic_text_oprimitive()", referenced from:
      dynet::save_dynet_model(std::basic_string<char, std::char_traits<char>, std::allocator<char> >, dynet::Model*) in model.cc.o
      boost::archive::text_oarchive_impl<boost::archive::text_oarchive>::~text_oarchive_impl() in model.cc.o
      boost::archive::text_oarchive_impl<boost::archive::text_oarchive>::~text_oarchive_impl() in model.cc.o
      boost::archive::text_oarchive::~text_oarchive() in model.cc.o
      boost::archive::text_oarchive::~text_oarchive() in model.cc.o
  "boost::archive::basic_binary_iprimitive<boost::archive::binary_iarchive, char, std::char_traits<char> >::init()", referenced from:
      dynet::VanillaLSTMBuilder::load_parameters_pretraining(std::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) in lstm.cc.o
      dynet::LSTMBuilder::load_parameters_pretraining(std::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) in lstm.cc.o
      dynet::SimpleRNNBuilder::load_parameters_pretraining(std::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) in rnn.cc.o
  "boost::archive::basic_binary_iprimitive<boost::archive::binary_iarchive, char, std::char_traits<char> >::load(std::basic_string<char, std::char_traits<char>, std::allocator<char> >&)", referenced from:
      boost::archive::detail::iserializer<boost::archive::binary_iarchive, std::pair<std::basic_string<char, std::char_traits<char>, std::allocator<char> > const, int> >::load_object_data(boost::archive::detail::basic_iarchive&, void*, unsigned int) const in dict.cc.o
      boost::archive::detail::iserializer<boost::archive::binary_iarchive, std::vector<std::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >::load_object_data(boost::archive::detail::basic_iarchive&, void*, unsigned int) const in dict.cc.o
      dynet::VanillaLSTMBuilder::load_parameters_pretraining(std::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) in lstm.cc.o
      dynet::LSTMBuilder::load_parameters_pretraining(std::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) in lstm.cc.o
      dynet::SimpleRNNBuilder::load_parameters_pretraining(std::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) in rnn.cc.o
  "boost::archive::basic_binary_iprimitive<boost::archive::binary_iarchive, char, std::char_traits<char> >::basic_binary_iprimitive(std::basic_streambuf<char, std::char_traits<char> >&, bool)", referenced from:
      dynet::VanillaLSTMBuilder::load_parameters_pretraining(std::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) in lstm.cc.o
      dynet::LSTMBuilder::load_parameters_pretraining(std::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) in lstm.cc.o
      dynet::SimpleRNNBuilder::load_parameters_pretraining(std::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) in rnn.cc.o
  "boost::archive::basic_binary_iprimitive<boost::archive::binary_iarchive, char, std::char_traits<char> >::~basic_binary_iprimitive()", referenced from:
      dynet::VanillaLSTMBuilder::load_parameters_pretraining(std::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) in lstm.cc.o
      dynet::LSTMBuilder::load_parameters_pretraining(std::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) in lstm.cc.o
      boost::archive::binary_iarchive_impl<boost::archive::binary_iarchive, char, std::char_traits<char> >::~binary_iarchive_impl() in lstm.cc.o
      boost::archive::binary_iarchive_impl<boost::archive::binary_iarchive, char, std::char_traits<char> >::~binary_iarchive_impl() in lstm.cc.o
      boost::archive::binary_iarchive::~binary_iarchive() in lstm.cc.o
      boost::archive::binary_iarchive::~binary_iarchive() in lstm.cc.o
      dynet::SimpleRNNBuilder::load_parameters_pretraining(std::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) in rnn.cc.o
      ...
  "boost::archive::basic_binary_oprimitive<boost::archive::binary_oarchive, char, std::char_traits<char> >::init()", referenced from:
      dynet::VanillaLSTMBuilder::save_parameters_pretraining(std::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const in lstm.cc.o
      dynet::LSTMBuilder::save_parameters_pretraining(std::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const in lstm.cc.o
      dynet::SimpleRNNBuilder::save_parameters_pretraining(std::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const in rnn.cc.o
  "boost::archive::basic_binary_oprimitive<boost::archive::binary_oarchive, char, std::char_traits<char> >::save(std::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)", referenced from:
      boost::archive::detail::oserializer<boost::archive::binary_oarchive, std::pair<std::basic_string<char, std::char_traits<char>, std::allocator<char> > const, int> >::save_object_data(boost::archive::detail::basic_oarchive&, void const*) const in dict.cc.o
      boost::archive::detail::oserializer<boost::archive::binary_oarchive, std::vector<std::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >::save_object_data(boost::archive::detail::basic_oarchive&, void const*) const in dict.cc.o
      dynet::VanillaLSTMBuilder::save_parameters_pretraining(std::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const in lstm.cc.o
      dynet::LSTMBuilder::save_parameters_pretraining(std::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const in lstm.cc.o
      boost::archive::detail::common_oarchive<boost::archive::binary_oarchive>::vsave(boost::archive::class_name_type const&) in lstm.cc.o
      dynet::SimpleRNNBuilder::save_parameters_pretraining(std::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const in rnn.cc.o
      boost::archive::detail::common_oarchive<boost::archive::binary_oarchive>::vsave(boost::archive::class_name_type const&) in rnn.cc.o
      ...
  "boost::archive::basic_binary_oprimitive<boost::archive::binary_oarchive, char, std::char_traits<char> >::basic_binary_oprimitive(std::basic_streambuf<char, std::char_traits<char> >&, bool)", referenced from:
      dynet::VanillaLSTMBuilder::save_parameters_pretraining(std::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const in lstm.cc.o
      dynet::LSTMBuilder::save_parameters_pretraining(std::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const in lstm.cc.o
      dynet::SimpleRNNBuilder::save_parameters_pretraining(std::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const in rnn.cc.o
  "boost::archive::basic_binary_oprimitive<boost::archive::binary_oarchive, char, std::char_traits<char> >::~basic_binary_oprimitive()", referenced from:
      dynet::VanillaLSTMBuilder::save_parameters_pretraining(std::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const in lstm.cc.o
      dynet::LSTMBuilder::save_parameters_pretraining(std::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const in lstm.cc.o
      boost::archive::binary_oarchive_impl<boost::archive::binary_oarchive, char, std::char_traits<char> >::~binary_oarchive_impl() in lstm.cc.o
      boost::archive::binary_oarchive_impl<boost::archive::binary_oarchive, char, std::char_traits<char> >::~binary_oarchive_impl() in lstm.cc.o
      boost::archive::binary_oarchive::~binary_oarchive() in lstm.cc.o
      boost::archive::binary_oarchive::~binary_oarchive() in lstm.cc.o
      dynet::SimpleRNNBuilder::save_parameters_pretraining(std::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const in rnn.cc.o
      ...
ld: symbol(s) not found for architecture x86_64
collect2: error: ld returned 1 exit status
make[2]: *** [dynet/libdynet.dylib] Error 1
make[1]: *** [dynet/CMakeFiles/dynet.dir/all] Error 2
make: *** [all] Error 2
```

**Comment by pmichel31415** [comment]
Is it possible you installed boost 32bit?

**Comment by yoavartzi** [comment]
I am using an Anaconda environment on a Mac, and it doesn't seem to support non 64bit installation: 
https://anaconda.org/anaconda/boost
I verified I have v1.61.0 installed

**Comment by geovedi** [comment]
How about using boost from Macports? I just tested building and running (some programs from `examples/cpp` directory) the latest commit from `master` without problem.

**Comment by yoavartzi** [comment]
I managed to install. I believe the problem is trying to use Conda's compiler rather than the one installed by xcode. I don't have the exact setup, but if you install all the prerequisites using Brew and then create a blank conda environment with only python and cython, compilation and installation seem to work. I had to install numpy afterwards to get it running, but it's likely I could have done it before.
